stan/math/opencl/matrix_cl.hpp

```diff
@@ -234,13 +232,13 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
    * @throw <code>std::invalid_argument</code> if the
    * matrices do not have matching dimensions
    */
-  template <int R, int C>
-  explicit matrix_cl(const std::vector<Eigen::Matrix<T, R, C>>& A) try
-      : rows_(A.empty() ? 0 : A[0].size()),
-        cols_(A.size()) {
+  template <typename Vec, require_std_vector_t<is_eigen, Vec>...>
```

Comment:
Containers of containers is a thing I need to work on. There's quite a few places where we have vectors of eigen matrices that then have a particular underlying scalar type. Not sure what is nicest syntax wise tho'

```
require_std_vector_contains_t<is_eigen, is_arithmetic, Vec>
```

I think I can do something like that

Reply:
Thinking about this, it may be better to do two requires here instead of having one joint on (otherwise combinatorally I think the actual code would start looking weird and jankey.) How do you feel about 
```
template <typename Vec, require_std_vector_vt<is_eigen, Vec>..., require_std_vector_st<is_arithmetic, Vec>...>
```

Where `_vt` would be equivalent to doing `value_type` while `_st` would be looking at `scalar_type`?

Reply:
That is fine with me. Thinking about it I have another idea that may be clearer and a bit more general. 

``` template<typename Vec, require_std_vector<Vec>, require_eigen<contained_t<Vec>>, require_arithmetic<contained_t<contained_t<Vec>>>```

`contained_t` is generalization for `std::vector<T>::value_type` or `Eigen::Matrix<...>::Scalar`.

If you prefere yours that is also fine with me.

Reply:
For yours we can use `value_type_t`. I have mixed feelings on the verbosity of it, though I think usually we either want to check

1. The containers underlying container (idt we have containers of containers of containers? Can we think of a situation where we explicitly have that?)
2. The containers underlying scalar type (var, fvar, arithmetic, yada yada)

I'll play around with both to see which is nicer